client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java

```diff
@@ -0,0 +1,279 @@
+            CompletableFuture<SessionHandler> r = connection.getSessionHandler();
+            return !r.isDone() || Futures.isSuccessful(r) && r.join().isConnectionEstablished();
+        }).collect(Collectors.toList());
+        log.debug("List of connections to {} that can be used: {}", location, prunedConnectionList);
+
+        // Fetch the Connection related stats.
+        ConnectionSummaryStats stats = prunedConnectionList.parallelStream().collect(collectorStats);
+        // Choose the connection with the least number of sessions.
+        Optional<Connection> suggestedConnection = stats.getConnectionWithMinimumSession(location);
+
+        final Connection connection;
+        if (suggestedConnection.isPresent() && stats.getConnectionCount(location) == clientConfig.getMaxConnectionsPerSegmentStore()) {
+            log.debug("Reusing connection: {}", suggestedConnection.get());
+            Connection oldConnection = suggestedConnection.get();
+            prunedConnectionList.remove(oldConnection);
+            connection = new Connection(oldConnection.getUri(), oldConnection.getSessionHandler(), oldConnection.getSessionCount() + 1);
+        } else {
+            // create a new connection.
+            log.debug("Creating a new connection to {}", location);
+            CompletableFuture<SessionHandler> sessionHandlerFuture = establishConnection(location);
+            connection = new Connection(location, sessionHandlerFuture, 1);
+        }
+        prunedConnectionList.add(connection);
+        connectionMap.put(location, prunedConnectionList);
+        return connection.getSessionHandler().thenApply(sessionHandler -> sessionHandler.createSession(session, rp));
+    }
+
+    @Override
+    public CompletableFuture<ClientConnection> getClientConnection(PravegaNodeUri location, ReplyProcessor rp) {
+        Preconditions.checkNotNull(location, "Location");
+        Preconditions.checkNotNull(rp, "ReplyProcessor");
+        Exceptions.checkNotClosed(closed.get(), this);
+
+        // create a new connection.
+        CompletableFuture<SessionHandler> sessionHandlerFuture = establishConnection(location);
+        Connection connection = new Connection(location, sessionHandlerFuture, 1);
+        return connection.getSessionHandler().thenApply(sessionHandler -> sessionHandler.createConnectionWithSessionDisabled(rp));
+    }
+
+    @Override
+    public int getActiveChannelCount() {
+        return (int) this.channelGroup.stream()
+                                      .filter(Channel::isActive)
+                                      .peek(ch -> log.debug("Channel with id {} localAddress {} and remoteAddress {} is active.", ch.id(),
+                                                            ch.localAddress(), ch.remoteAddress()))
+                                      .count();
+    }
+
+    @VisibleForTesting
+    public List<Channel> getActiveChannels() {
```

Comment:
This method is awfully similar to the one above. Can you consolidate?

Reply:
done.